### ISSUE TITLE
Change default built-in terminal colors

### ISSUE THREAD

**Issue by echasnovski** [neovim/neovim#26857]
### Problem

After #26540, default color scheme uses custom palette when true colors are enabled. The exception is colors in `:terminal`, which use palette from terminal emulator.

### Expected behavior

Use built-in palette by default.

Essentially, setting default terminal colors similar to the following Lua code:

```lua
local bg = vim.o.background == 'dark' and 'NvimDark' or 'NvimLight'
local fg = bg == 'NvimDark' and 'NvimLight' or 'NvimDark'

vim.g.terminal_color_0  = vim.g.terminal_color_0  or (bg .. 'Grey2')
vim.g.terminal_color_1  = vim.g.terminal_color_1  or (fg .. 'Red')
vim.g.terminal_color_2  = vim.g.terminal_color_2  or (fg .. 'Green')
vim.g.terminal_color_3  = vim.g.terminal_color_3  or (fg .. 'Yellow')
vim.g.terminal_color_4  = vim.g.terminal_color_4  or (fg .. 'Blue')
vim.g.terminal_color_5  = vim.g.terminal_color_5  or (fg .. 'Magenta')
vim.g.terminal_color_6  = vim.g.terminal_color_6  or (fg .. 'Cyan')
vim.g.terminal_color_7  = vim.g.terminal_color_7  or (fg .. 'Grey2')
vim.g.terminal_color_8  = vim.g.terminal_color_8  or (bg .. 'Grey2')
vim.g.terminal_color_9  = vim.g.terminal_color_9  or (fg .. 'Red')
vim.g.terminal_color_10 = vim.g.terminal_color_10 or (fg .. 'Green')
vim.g.terminal_color_11 = vim.g.terminal_color_11 or (fg .. 'Yellow')
vim.g.terminal_color_12 = vim.g.terminal_color_12 or (fg .. 'Blue')
vim.g.terminal_color_13 = vim.g.terminal_color_13 or (fg .. 'Magenta')
vim.g.terminal_color_14 = vim.g.terminal_color_14 or (fg .. 'Cyan')
vim.g.terminal_color_15 = vim.g.terminal_color_15 or (fg .. 'Grey2')
```

**Comment by clason** [comment]
I think I would prefer my `:terminal` to look the same as the host terminal, colors included.

**Comment by echasnovski** [comment]
I can see that this is debatable, hence an issue first without an actual PR.

My reasoning here is that `:terminal` uses background and foreground from `Normal` and not from terminal emulator. As default color scheme defines its own `Normal` (which might be different from that of terminal emulator), to me a more consistent behavior would be for `:terminal`to also use colors from Neovim's default color scheme.

**Comment by clason** [comment]
> As default color scheme defines its own Normal (which might be different from that of terminal emulator),

Not sufficiently so; that's the point of `bg=dark` or `light`.

**Comment by clason** [comment]
I think this would make sense as an _option_ (not in the literal sense, but a script users could source in their `init.lua` if the want full Neovim branding). My main argument for not doing this by default is that it's (much) less effort to opt-in than to opt-out.

**Comment by echasnovski** [comment]
I see. Well, it won't be a default then. This is the same line of reasoning as in "Why change default color scheme if it is better to create 'runtime/colors/neovim.lua' and one can just `:color neovim`?".

Do you think more as a code snippet in documentation (for example in 'nvim_terminal_emulator.txt' or in 'syntax.txt' near `*gui-colors*`) or as some custom script (plus documentation)?

**Comment by clason** [comment]
I was thinking custom script (placing tbd). 

> This is the same line of reasoning as in "Why change default color scheme if it is better to create 'runtime/colors/neovim.lua' and one can just :color neovim?".

Because some colors is better than no colors; and if you have to set some colors, they should be usable. This here is crossing into "personal preference" territory. So I don't think that's comparable at all.

**Comment by glacambre** [comment]
Two cents from a GUI maintainer: at the moment, neovim sends 224 (that's `#0000e0` I think) as foreground color for "terminal color 4" to GUIs. In my opinion, it doesn't mesh well with the default color scheme and would be nicer if a color from the default color scheme was used instead. I don't know if this falls under the scope of this issue though.

**Comment by craigmac** [comment]
Came here after realizing `:term` in Neovide build doesn't match the default colorscheme. I think for GUIs especially it should be set to, or near to, the new default colorscheme.

**Comment by echasnovski** [comment]
@justinmk, do you mind me asking about the executive decision here? Basically, should colors for built-in terminal (`:terminal`) be inherited from default color scheme or left as current is as terminal emulator's colors?

If the latter, then should there be some kind of opt-in script for "enhancing" experience with default color scheme? It could define terminal colors and maybe some more. If yes, what is the place to put it? There is 'scripts' directory, but not sure if it is appropriate to expose for end users.

Another idea is that the terminal colors can be defined in 'runtime/colors/default.vim'. In this case `nvim --clean` would use terminal emulator's palette, but `:colorscheme default` would define a more aligned palette.

**Comment by justinmk** [comment]
@echasnovski similar to the current upvotes/downvotes count, I'm on the fence about this. And when something is a coin-flip, we usually choose to "do nothing". 


> should there be some kind of opt-in script for "enhancing" experience with default color scheme?

We could start with that. It could live in `runtime/pack/dist/opt/`

**Comment by clason** [comment]
Conversely, the new default colorscheme is now [widely available](https://github.com/mbadolato/iTerm2-Color-Schemes?tab=readme-ov-file#nvimdark) as a terminal colorscheme.

**Comment by craigmac** [comment]
The default terminal colors when using `:terminal` in Neovide (latest, 0.13) are barely useable. This is not just a personal preference but an accessibility concern, which was one of the goals of the new default colorscheme if I'm not mistaken. Can we support both cases presented here by only setting these when GUI is attached, by guarding it with a `vim.fn.has('gui_running')`? This way GUIs get the branded version out of the box, and non-gui users aren't affected.

**Comment by clason** [comment]
Can you back that up with a screenshot?

**Comment by lsvmello** [comment]
Most builtin colorschemes set terminal colors if `termguicolors == true` or `has('gui_running')`:

https://github.com/neovim/neovim/blob/e71d2c817d1a2475551f58a98e411f6b39a5be3f/runtime/colors/elflord.vim#L50-L56

Why shouldn't `runtime/colors/default.vim` do the same? The caveat is that user should explicitly `:colorscheme default`, but that could be handled by GUIs _(check if colorscheme is not set and set `default`)_.

**Comment by bfredl** [comment]
We might want to update the default `guifg` `guibg` values that we transmit to GUI:s not using indexed colors.
But that does not imply changing the behavior of the builtin TUI with 'termguicolors' which still should be passing through colors.

**Comment by fredizzimo** [comment]
> The caveat is that user should explicitly :colorscheme default, but that could be handled by GUIs (check if colorscheme is not set and set default). 

GUIs currently have no hook for doing that https://github.com/neovim/neovim/issues/25377. They can set options before attaching, on `VimEnter` or on `UIEnter` both are called after the potential first render of the screen, so it would cause flickering.

But otherwise I agree, at least for GUIs the default should be defined by the colorscheme.

**Comment by lsvmello** [comment]
> GUIs currently have no hook for doing that https://github.com/neovim/neovim/issues/25377. They can set options before attaching, on `VimEnter` or on `UIEnter` both are called after the potential first render of the screen, so it would cause flickering.

I don't think it would flicker because `colors/default.vim` does not set any highlights, but it would clear out custom user highlights if not set before user's configuration.

**Comment by fredizzimo** [comment]
> I don't think it would flicker because colors/default.vim does not set any highlights, but it would clear out custom user highlights if not set before user's configuration.

If it needs to be done before any user configuration, then I'm afraid that it is completely impossible to do from the GUI currently. If we call it before `nvim_ui_attach`, then `has('gui_running')` will be false, and if called after that, then the init file will already be loading. For attaching to remote servers it gets even more tricky.  So the best way to take care of this is from Nvim itself.

For embedded mode it might be as simple as moving the `init_highlight` call after the remote UI wait. For remote UIs, see this https://github.com/neovim/neovim/issues/29634.

**Comment by jmpolom** [comment]
> And when something is a coin-flip, we usually choose to "do nothing".

You claim this project management method is _only_ used when the devs perceive 2 competing courses of action as equivalent? Are you sure??
